snapcraft/plugins/ruby.py

```diff
@@ -0,0 +1,157 @@
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""The ruby plugin is useful for ruby based parts.
+The plugin uses gem to install dependencies from a `Gemfile`.
+This plugin uses the common plugin keywords as well as those for "sources".
+For more information check the 'plugins' topic for the former and the
+'sources' topic for the latter.
+Additionally, this plugin uses the following plugin-specific keywords:
+    - gems:
+      (list)
+      A list of gems to install.
+    - use-bundler
+      (boolean)
+      Use bundler to install gems from a Gemfile (defaults 'false').
+    - ruby-version:
+      (string)
+      The version of ruby you want this snap to run.
+"""
+import logging
+import os
+import platform
+import re
+
+from snapcraft import BasePlugin, file_utils
+from snapcraft.sources import Tar
+
+
+logger = logging.getLogger(__name__)
+
+
+class RubyPlugin(BasePlugin):
+
+    @classmethod
+    def schema(cls):
+        schema = super().schema()
+
+        schema['properties']['use-bundler'] = {
+            'type': 'boolean',
+            'default': False
+        }
+        schema['properties']['ruby-version'] = {
+            'type': 'string',
+            'default': '2.4.0'
+        }
+        schema['properties']['gems'] = {
+            'type': 'array',
+            'minitems': 1,
+            'uniqueItems': True,
+            'items': {
+                'type': 'string'
+            },
+            'default': []
+        }
+        return schema
+
+    @classmethod
+    def get_pull_properties(cls):
+        # Inform Snapcraft of the properties associated with pulling. If these
+        # change in the YAML Snapcraft will consider the build step dirty.
+        return ['ruby-version', 'gems', 'use-bundler']
+
+    def __init__(self, name, options, project):
+        super().__init__(name, options, project)
+        # Beta Warning
+        # Remove this comment and warning once ruby plugin is stable.
+        logger.warn("The ruby plugin is currently in beta, "
+                    "its API may break. Use at your own risk")
+
+        self._ruby_version = options.ruby_version
+        self._ruby_part_dir = os.path.join(self.partdir, 'ruby')
+        self._ruby_download_url = \
+            'https://cache.ruby-lang.org/pub/ruby/ruby-{}.tar.gz'.format(
+                self._ruby_version)
+        self._ruby_tar = Tar(self._ruby_download_url, self._ruby_part_dir)
+        self._gems = options.gems or []
+        self._ruby_version_dir = '{}.{}.0'.format(
```

Comment:
It's still pretty hacky. I was okay waiting until later, but if you want to re-design this bit now I've actually pushed a commit here to show you what I had in mind. It uses globbing to discover the version, so no programatic trickery is required. Feel free to delete the commit if you dislike it.